### ISSUE TITLE
Recursion almost broken

### ISSUE THREAD

**Issue by hryamzik** [ansible/ansible#16749]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.0.0
```
##### OS / ENVIRONMENT

Mac OS
##### SUMMARY

Recursion includes gets slower and slower till become useless.
##### STEPS TO REPRODUCE

``` yml
- hosts: localhost
  become: no
  become_method: sudo
  vars:
      i: 0
  tasks:
      - include: include_me.yml

      - debug: msg="Exited, i={{ i }}"
```

include_me.yml:

``` yml
- debug: var=i

- set_fact:
    i: "{{ (i|int) + 1 }}"

- include: include_me.yml
  when: "{{ i|int < 30 }}"
```

Playbook run took 0 days, 1 hours, 38 minutes, 3637 seconds

This simple sequence takes forever to execute on my retina MBP. In my real life scenario hundreds of iterations are used to download a full backup with all the incrementals so this issue breaks the whole idea of searching for previous backup with ansible.

**Comment by jimi-c** [comment]
So testing out my performance branch (https://github.com/ansible/ansible/compare/performance_improvements), and upping the level of recursion even to 200 levels deep results in an execution time of about 33 seconds:

<pre>
$ time ansible-playbook -vv test.yml -e max_depth=200
...
PLAY RECAP *********************************************************************
localhost                  : ok=601  changed=0    unreachable=0    failed=0   
real    0m33.785s
user    0m24.153s
sys    0m6.773s
</pre>

If you'd like to help test out that branch, I'd appreciate it!

**Comment by jeremyeder** [comment]
James -- hopefully you saw the results I sent you by email.

**Comment by jimi-c** [comment]
@jeremyeder I did, those were using this performance branch? I was under the impression that those were with restructured playbooks rather than a different branch.

**Comment by hryamzik** [comment]
Is there any ETA for getting this in released version?

**Comment by jeremyeder** [comment]
Those were using both restructured playbooks and your branch.

**Comment by jimi-c** [comment]
@hryamzik the goal is to get this into 2.2, which is aiming for a September release. Since this patch introduces some fairly major restructuring internally, I'd really like some additional testing on it to make sure we have a good level of confidence that nothing was horribly broken before merging it into devel.

@jeremyeder excellent, I didn't realize Andrew had gotten past some of the early bugs in that branch. There was a lot of work later in the day yesterday, so I'd appreciate it if you re-ran those tests with the most recent version as well.

**Comment by jeremyeder** [comment]
Will do...but not until end of next week /cc @mffiedler @timothysc @abutcher

**Comment by jeremyeder** [comment]
Here is the resource usage of ansible2.2-performance_improvements branch.  To deploy a 100-node openshift cluster, the install took 21m48s.

3 masters
3 etcd
1 apiserver load balancer
2 infrastructure nodes
100 compute nodes

![ansible22-pi](https://cloud.githubusercontent.com/assets/1919640/17552444/0278cff0-5ece-11e6-85bf-fec9471b7811.png)

Overall this is a significant improvement over ansible-2.1 where the same work would not complete given 64GB of RAM and over 8 hours of runtime before OOM.

 /cc @mffiedler @timothysc @abutcher

**Comment by jimi-c** [comment]
@jeremyeder it looks like it shaved about 20 minutes off the execution time from the previous results you emailed me?

**Comment by jimi-c** [comment]
As the performance improvements branch has been merged into devel now, and as the deeply nested include speed issue has been resolved, we're going to go ahead and close this issue now. These improvements will be included in 2.2.0 (currently targeted for release in September of 2016).

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
